sidestep_engine/core/adaptive_timestep.py

```
@@ -0,0 +1,176 @@
+        ema_decay: Smoothing factor for per-bin loss tracking.
+        ratio: Fraction of batch sampled adaptively (rest from base).
+    """
+
+    def __init__(
+        self,
+        n_bins: int = 10,
+        ema_decay: float = 0.99,
+        ratio: float = 0.3,
+    ) -> None:
+        if not (0.0 <= ratio <= 1.0):
+            raise ValueError(f"ratio must be in [0, 1] (got {ratio})")
```

Ruff (0.15.2)

[warning] Avoid specifying long messages outside the exception class

(TRY003)